Add Two-way binding between signals and query params post (#9)
Zuehlke/software-engineering-corner: drafts/two-way-binding-signals-and-query-params.md

  selector: "app-root",
  standalone: true,
  template: ` <h1>Signals Demo</h1> `,
})
export class AppComponent {}
```

![Initial application](https://cdn.hashnode.com/res/hashnode/image/upload/v1710536390017/q5rnF-2Oc.png?auto=format&w=500)

This is pretty much what you get when you generate a new Angular project nowadays. But without all the boilerplate and some basic styles.

Let's start by consuming query params in the traditional way, using RxJS:

```ts
@Component({
  selector: "app-root",
  standalone: true,
  imports: [AsyncPipe],
  template: `
    <h1>Signals Demo</h1>
    <p>Your first name is: {{ firstName$ | async }}</p>
  `,
})
export class AppComponent {
  private activatedRoute = inject(ActivatedRoute);

  firstName$ = this.activatedRoute.queryParams.pipe(
    map((allQueryParams) => allQueryParams["firstName"])
  );
}
```

![Consuming query params with RxJS](https://cdn.hashnode.com/res/hashnode/image/upload/v1710536464769/FbnjIrApA.gif?auto=format)

Great! This works and it's not necessarily inconvenient since Angular is handling the tedious task of managing subscriptions when we use the `async` pipe. Now let's see how this would look like with signals.

> At the moment of writing this article, Angular's router doesn't provide signals in its API. They will most likely get introduced in future releases, but for now we have to do a workaround. We can pick the current responses from Angular's router, which are based on RxJS's observables, and convert them by using the `toSignal` utility function:

Reviewer: at the time of writing this

Reviewer: and convert them to Signals by using the toSignal utility
idk now you have Singal twice, but imo this is clearer

```ts

> @Component({

Reviewer: how many imports are there? maybe it makes sense to add them, or just to add the one which you added here like the toSingal import

Let's make the example a bit more interesting by adding a few more query params and extracting the logic of creating the signal to a private method:

```ts
@Component({
  selector: "app-root",
  standalone: true,
  template: `
    <h1>Signals Demo</h1>
    <p>Full name: {{ fullName() }}</p>
    <p>Age: {{ age() }}</p>
  `,
})
export class AppComponent {
  private activatedRoute = inject(ActivatedRoute);

  firstName = this.getQueryParam("firstName");
  lastName = this.getQueryParam("lastName");
  age = this.getQueryParam("age");

  fullName = computed(() => `${this.firstName() ?? ""} ${this.lastName() ?? ""}`);

  private getQueryParam(name: string) {
    const value$ = this.activatedRoute.queryParams.pipe(
      map((allQueryParams) => allQueryParams[name])
    );

    return toSignal(value$);
  }
}
```

![Consuming query params with signals](https://cdn.hashnode.com/res/hashnode/image/upload/v1710536530242/EJ6p_x7xp.gif?auto=format)

> This is already pretty cool! We have a handy utility to consume query params. Notice how easily we combined the `firstName` and `lastName` signals to compute the `fullName`, this would have been much more verbose and complicated using RxJS.

Reviewer: handy utility function

Finally, the UI gets updated successfully because, once the router finishes the navigation, a new value will be emitted in the `activatedRoute.queryParams` observable, which we map and convert to the signals that we have in our component. So every time it emits a new value, our signals will be in sync.

This is very cool but we can do better:

```ts
anonymise() {
  // It's not very nice that we have to specify the query params again here ⬇️
  const queryParams = { firstName: undefined, lastName: undefined };
  this.router.navigate([], { queryParams, queryParamsHandling: "merge" });
}
```

Notice that we already have the `firstName` and `lastName` signals in our component. Wouldn't it be nicer if we could just **set** the new values of the query params on the signals directly? Something like this:

```ts
anonymise() {
  this.firstName.set(undefined)
  this.lastName.set(undefined)
}
```

That would be amazing, but unfortunately this doesn't work since the `toSignal` utility returns a read-only signal.

That doesn't mean we cannot do it ourselves though. Let's build it!

## Implementing Two-way binding between Signals and Query Params

> First let's create a handy `QueryParamsService` so we separate the query param handling logic from our component:

Reviewer: i would remove handy here :)

```ts
@Injectable({ providedIn: "root" })
export class QueryParamService {
  private activatedRoute = inject(ActivatedRoute);

  get(queryParamName: string) {
    const value$ = this.activatedRoute.queryParams.pipe(
      map((allQueryParams) => allQueryParams[queryParamName])
    );

    return toSignal(value$);
  }
}
```

```ts
@Component({
  selector: "app-root",
  standalone: true,
  template: `
    <h1>Signals Demo</h1>
    <p>Full name: {{ fullName() }}</p>
    <p>Age: {{ age() }}</p>

    <button (click)="anonymise()">I want to be anonymous 👤</button>
  `,
})
export class AppComponent {
  private queryParamService = inject(QueryParamService);

  firstName = this.queryParamService.get("firstName");
  lastName = this.queryParamService.get("lastName");
  age = this.queryParamService.get("age");

  fullName = computed(() => `${this.firstName() ?? ""} ${this.lastName() ?? ""}`);

  anonymise() {
    // TODO: in a minute :)
  }
}
```

> This small refactor will produce the same behavior as before except for the anonymising feature, which we will implement more adequately as soon as we make our `get` method from `QueryParamService` return a writable signal.

Reviewer: i have no idea about commas in english, but for me it looks better if there is a comma before except and not before which
but tbh i have no clue :D

```ts
@Injectable({ providedIn: "root" })
export class QueryParamService {
  private activatedRoute = inject(ActivatedRoute);
  private allQueryParams = toSignal(this.activatedRoute.queryParams);

  get(queryParamName: string) {
    const queryParamValue = signal("");

    effect(
      () => {
        const newQueryParamValue = this.allQueryParams()?.[queryParamName];
        queryParamValue.set(newQueryParamValue);
      },
      { allowSignalWrites: true }
    );

    return queryParamValue;
  }
}
```

Since we are accessing the value of the `allQueryParams` signal in the effect, it will run every time this signal gets updated, which happens every time Angular emits a new value in the `activatedRoute.queryParams` observable.

Inside the effect, we are just updating the value of our `queryParamValue` signal. For that, notice that we need to pass the `allowSignalWrites: true` option. This is necessary because updating signals in effects could lead to infinite loops and unexpected and intricate situations in general. That's why Angular disables this possibility by default.

> However, in this scenario there's no danger to do it: The effect is only triggered when `allQueryParams` is updated, that refreshes our `queryParamValue` signal and it finishes there. No danger of triggering an infinite loop or unexpected situations.

Reviewer: there's no danger to it

```typescript
const activatedRouteValue = signal("red");
const queryParamValue = signal("blue");

// Something triggers a change in one signal. For example:
activatedRouteValue.set("purple");

// activatedRouteValue has changed from "red" to "purple" then trigger first effect:

// (Inside activatedRouteValue->queryParamValue effect):
queryParamValue.set("purple");

// queryParamValue has changed from "blue" to "purple" then trigger second effect:

// (Inside queryParamValue->activatedRouteValue effect):
activatedRouteValue.set("purple");

// activatedRouteValue is already "purple". It hasn't changed its value. Therefore,
// there's no need to trigger first effect again. End of the chain.
```

So this should be it, right? Well, unfortunately this doesn't work yet:

![Empty UI situation when using our almost ready solution](https://cdn.hashnode.com/res/hashnode/image/upload/v1710538597922/uDBmH5mo6.gif?auto=format)

Apparently we have an empty state. Not only that, query params disappear from the browser URL. What is going on?

> Well, most of it is related to Angular's router start up logic. Initially, Angular only emits the actual or real values of the query params through the `activatedRoute.queryParams` observable **when the first navigation** is completed. 

Reviewer: startup

This sounds surprising at first, specially when we try to access a query param that is initially provided in the browser URL and, through Angular's router API, its starting value is `undefined`. Why doesn't Angular just pick whatever is initially in the browser URL?

If you think about it, this behavior makes sense, there could be logic configured that affects the initial and all navigations in general, `guards` could be a good example of that. This means, some query params could be added or removed during a navigation, not necessarily matching what was initially available in the browser URL. And because of this, Angular is only sure about the actual or correct values of the query params as soon as the current navigation finishes.

> During start up, no navigation has been completed yet. The current query params object in the Angular router is empty `{}`. If we try to access any query param in this period of time, it will be `undefined`. This situation affects us in the initial value we provide to our `queryParamValue` signal. If we use the `get` method from the `QueryParamService` in components that are loaded during the initial navigation, then our `queryParamValue` signal will be initialized to `undefined`. This is technically not a big deal since it is connected to the `activatedRoute.queryParams` observable. So as soon as Angular emits the actual value of the query param, then the signal will be updated accordingly. But the problem would have already occurred by then.

Reviewer: startup

---
title: Two-way binding between Signals and Query Params
subtitle: "Pushing the limits of signals while rediscovering Angular concepts"
domain: software-engineering-corner.zuehlke.com
tags: signals, web-development, javascript, typescript, frontend-development, reactivity, explained, angular
cover: https://cdn.hashnode.com/res/hashnode/image/upload/v1710531988213/wpYOexURa.jpg?auto=format
publishAs: juliocastrodev
hideFromHashnodeCommunity: false
saveAsDraft: true
---

> Signals are the hot topic of today. Pretty much every major framework (except `React`...) is introducing them in one way or another. This is surprisingly also the case of Angular, which is making huge efforts to improve a lot of the pain points we've been experiencing for too many years. A few examples of this include standalone components, a new build system leveraging esbuild and our protagonist of today: signals. 

Reviewer 2: **nitpick:** I wouldn't make a framework/library name in "code style" ;) same with "RxJS" in the rest of the article
```suggestion
Signals are the hot topic of today. Pretty much every major framework (except React...) is introducing them in one way or another. This is surprisingly also the case of Angular, which is making huge efforts to improve a lot of the pain points we've been experiencing for too many years. A few examples of this include standalone components, a new build system leveraging esbuild and our protagonist of today: signals. 
```

If you want to learn more about what signals are and a few interesting historical details about them, reach out to our amazing previous article [Signals, Signals everywhere][signalsEverywhereArticle] by [Lucas Schnüriger][lucasProfile].

> The whole concept of signals is a pretty new topic in the Angular community. In fact, just a few weeks ago, [Angular 17.2][angular17.2] was released, introducing the possibility of using two-way binding along with signals. And not long before, [Angular 17.1][angular17.1] released a new API to specify `@Input()` while working with results as signals. 

Reviewer 2: **question:** What does the last sentence mean? I don't quite understand "working with results as signals"?

Author: Yeah, I'll probably rephrase it. I wanted to say that using  the new `input` api, Angular provides the result of the input (what the parent is sending to the child) as a signal. So if parent is sending `100` then the children will get a `Signal<number>` with that value.

> Wait a minute, really? Weren't those important APIs already available when the Angular team released the first stable version of signals? Isn't two-way binding or, even more, parent-to-child communication basic features of the framework? Yes, we could claim that but the Angular team is taking a gradual strategy to introduce all these new concepts into the framework. And that's probably a good thing, since they are learning and listening from the community during the process. You can also influence and give your opinion in the [open discussions][angularDiscussions].

Reviewer 2: ```suggestion
Wait a minute, really? Weren't those important APIs already available when the Angular team released the first stable version of signals? Isn't two-way binding or, even more, parent-to-child communication basic features of the framework? Yes, we could claim that but the Angular team is taking a gradual strategy to introduce all these new concepts into the framework. And that's probably a good thing, since they are learning and listening from the community during the process. You can also influence it and give your opinion in the [open discussions][angularDiscussions].
```

> In today's article, we'll experiment and push to the limit the current state of signals in the Angular framework. For that, we will implement an exciting Two-way binding mechanism to connect signals with Angular's router state and the browser url. Are you ready? 

Reviewer 2: ```suggestion
In today's article, we'll experiment and push the limits of the current state of signals in the Angular framework. For that, we will implement an exciting two-way binding mechanism to connect signals with Angular's router state and the browser url. Are you ready? 
```

Author: I'll use Grammarly next time 😆

## The Challenge

> Let's start with a simple application:

Reviewer: i agree:
you can still say lets start with a simple application, which you can generate via the CLI. But imo you could save the code example and screenshot

```ts
@Component({
  selector: "app-root",
  standalone: true,
  template: ` <h1>Signals Demo</h1> `,
})
export class AppComponent {}
```

![Initial application](https://cdn.hashnode.com/res/hashnode/image/upload/v1710536390017/q5rnF-2Oc.png?auto=format&w=500)

This is pretty much what you get when you generate a new Angular application nowadays. But without all the boilerplate and some basic styles.

Let's start by consuming query params in the traditional way: using `RxJS`:

```ts
@Component({
  selector: "app-root",
  standalone: true,
  imports: [AsyncPipe],
  template: `
    <h1>Signals Demo</h1>
    <p>Your first name is: {{ firstName$ | async }}</p>
  `,
})
export class AppComponent {
  private activatedRoute = inject(ActivatedRoute);

  firstName$ = this.activatedRoute.queryParams.pipe(
    map((allQueryParams) => allQueryParams["firstName"])
  );
}
```

![Consuming query params with RxJS](https://cdn.hashnode.com/res/hashnode/image/upload/v1710536464769/FbnjIrApA.gif?auto=format)

> Great! this works and it's not necessarily inconvenient since Angular is handling the tedious task of managing subscriptions when we use the `async` pipe. Now let's see how this would look like with signals:

Reviewer 2: ```suggestion
Great! This works and it's not necessarily inconvenient since Angular is handling the tedious task of managing subscriptions when we use the `async` pipe. Now let's see how this would look like with signals.
```

> At the moment of writing this article, Angular's router doesn't provide signals as results when we use its API. Most likely this will be introduced in future releases, but for now we have to do a workaround. We can pick the current responses from Angular's router, which are based on `RxJS`'s observables, and convert them by using the `toSignal` utility function:

Reviewer 2: ```suggestion
At the moment of writing this article, Angular's router doesn't provide signals in its API. They will most likely get introduced in future releases, but for now we have to do a workaround. We can pick the current responses from Angular's router, which are based on RxJS's observables, and convert them by using the `toSignal` utility function:
```

```ts
@Component({
  selector: "app-root",
  standalone: true,
  template: `
    <h1>Signals Demo</h1>
    <p>Your first name is: {{ firstName() }}</p>
  `,
})
export class AppComponent {
  private activatedRoute = inject(ActivatedRoute);

  firstName = toSignal(

>     this.activatedRoute.queryParams.pipe(map((allQueryParams) => allQueryParams["firstName"]))

Reviewer 2: **suggestion:** probably makes sense to put `map` on a new line again (like in the previous example), code in Hashnode is quite large and let's avoid horizontal scrolling as much as possible

  private router = inject(Router);

  // ...

  anonymise() {
    const queryParams = { firstName: undefined, lastName: undefined };
    this.router.navigate([], { queryParams, queryParamsHandling: "merge" });
  }

  // ...
}
```

![Changing query params with Angular router](https://cdn.hashnode.com/res/hashnode/image/upload/v1710536580024/nqZ9dSNZx.gif?auto=format)

That's amazing! everything is working as expected, our signals are correctly getting the new `undefined` values and Angular is updating the UI accordingly.

Let's take a closer look in this `anonymise` method: We are using Angular's router to navigate to the exact same page, that's the reason we are passing an empty array:

```ts
this.router.navigate([]);
```

We do this because we are only interested to change the values of the query params, not the route or anything else.

In this case we want to unset the `firstName` and `lastName` query params, so we can anonymise the user. That's why we are passing `undefined` as their new values:

```ts
const queryParams = { firstName: undefined, lastName: undefined };
this.router.navigate([], { queryParams, queryParamsHandling: "merge" });
```

> The `queryParamsHandling: "merge"` option is basically telling Angular to perform an **update** of the query params. This is important because, in this case, we are not interested to lose the `age` query param whenever it's there.

Reviewer 2: **thought:** I think the whole section from line 154 to here could be reduced. Changing query params in Angular is not a remarkably unique or complicated task. Thinking about the target audience of this article, I would assume them to at least understand the code above and a very short summary (without repeating the snippets) would help everyone else enough.

Finally, the UI gets updated successfully because after the router navigates, a new value will be emitted in the `activatedRoute.queryParams` observable, which we map and convert to the signals that we have in our component. So every time it emits a new value, our signals will be in sync.

This is very cool but we can do better:

```ts
anonymise() {
  // It's not very nice that we have to specify the query params again here ⬇️
  const queryParams = { firstName: undefined, lastName: undefined };
  this.router.navigate([], { queryParams, queryParamsHandling: "merge" });
}
```

> Notice that we already have in our component a `firstName` and `lastName` signals. Wouldn't be nicer if we can just **set** the new values of the query params by using the signals directly? Something like this:

Reviewer 2: ```suggestion
Notice that we already have the `firstName` and `lastName` signals in our component. Wouldn't it be nicer if we could just **set** the new values of the query params on the signals directly? Something like this:
```

```ts
anonymise() {
  this.firstName.set(undefined)
  this.lastName.set(undefined)
}
```

That would be amazing, but unfortunately this doesn't work since the `toSignal` utility returns a read-only signal.

> But that doesn't mean we cannot do it ourselves. Let's build it!

Reviewer 2: **suggestion:** rephrase to avoid "but …" twice in short succession

More specifically, we need to connect it to the `activatedRoute.queryParams` observable, but manually creating and maintaining a subscription in our service doesn't sound like a signal-friendly idea:

```ts
@Injectable({ providedIn: "root" })
export class QueryParamService {
  private activatedRoute = inject(ActivatedRoute);

  get(queryParamName: string) {
    const queryParamValue = signal("");

    // This will "work" but we don't want to explicitly subscribe here ❌
    this.activatedRoute.queryParams.subscribe((allQueryParams) =>
      queryParamValue.set(allQueryParams[queryParamName])
    );

    return queryParamValue;
  }
}
```

A signal-friendly idea could be to convert the `activatedRoute.queryParams` observable to a signal itself and then somehow synchronize it with `queryParamValue`:

```ts
@Injectable({ providedIn: "root" })
export class QueryParamService {
  private activatedRoute = inject(ActivatedRoute);
  private allQueryParams = toSignal(this.activatedRoute.queryParams);

  get(queryParamName: string) {
    const queryParamValue = signal("");

    // TODO: sync queryParamValue signal with allQueryParams signal

    return queryParamValue;
  }
}

> ```

Reviewer 2: **suggestion:** skip this code, it doesn't add much here, just have the two paragraphs and then code. For an article, too many code snippets will make it longer and harder to read. Readers will eventually skip, if they can't pinpoint the exact changes quickly.

        const newQueryParamValue = this.allQueryParams()?.[queryParamName];
        queryParamValue.set(newQueryParamValue);
      },
      { allowSignalWrites: true }
    );

    return queryParamValue;
  }
}
```

Since we are accessing the value of the `allQueryParams` signal in the effect, it will run every time this signal gets updated, which happens every time Angular emits a new value in the `activatedRoute.queryParams` observable.

Inside the effect, we are just updating the value of our `queryParamValue` signal. For that, notice that we need to pass the `allowSignalWrites: true` option. This is necessary because updating signals in effects could lead to infinite loops and unexpected and intricate situations in general. That's why Angular disables this possibility by default.

However, in this scenario there's no danger to do it: The effect is only triggered when `allQueryParams` is updated, that refreshes our `queryParamValue` signal and it finishes there. No danger of triggering an infinite loop or unexpected situations.

![Using our partial solution to access query params](https://cdn.hashnode.com/res/hashnode/image/upload/v1710538037690/Nk-9yMBSk.gif?auto=format)

> With the current solution we achieve the same behavior as before. However, we are not yet able to implement the anonymise feature in the nicer way. Yes, now we are returning a writable signal but if we set or update its value there's going to be no effect in Angular's router state or in the browser url. We'll achieve that in a minute but before that let's improve a bit the current solution.

Reviewer 2: ```suggestion
With the current solution we achieve the same behavior as before. However, we are not yet able to implement an improved anonymise feature. Yes, now we are returning a writable signal but if we set or update its value there's going to be no effect in Angular's router state or in the browser url. We'll achieve that in a minute but first let's improve a bit the current solution.
```

// queryParamValue has changed from "blue" to "purple" then trigger second effect:

// (Inside queryParamValue->activatedRouteValue effect):
activatedRouteValue.set("purple");

// activatedRouteValue is already "purple". It hasn't changed its value. Therefore,
// there's no need to trigger first effect again. End of the chain.
```

So this should be it, right? Well, unfortunately this doesn't work yet:

![Empty UI situation when using our almost ready solution](https://cdn.hashnode.com/res/hashnode/image/upload/v1710538597922/uDBmH5mo6.gif?auto=format)

Apparently we have an empty state. Not only that, query params disappear from the browser url. What is going on?

Well, most of it is related to Angular's router start up logic. Initially, Angular only emits the actual or real values of the query params through the `activatedRoute.queryParams` observable **when the first navigation** is completed. 

This sounds surprising at first, specially when we try to access a query param that is initially provided in the browser url and, through Angular's router API, its starting value is `undefined`. Why doesn't Angular just pick whatever is initially in the browser url?

> To understand it better let's consider an example. Imagine that you visit `http://localhost:4200/home?firstName=Dave` and, in the component assigned to `/home`, you try to access `activatedRoute.snapshot.queryParams["firstName"]`, in its `ngOnInit` for example. Angular won't return `Dave` as you may expect. You will actually receive `undefined`. The reason is that Angular hasn't finished the initial navigation yet. During this period of time, if we access the `snapshot` or the current value of the `activatedRoute.queryParams` observable, we will get an empty object `{}`, that's why we received `undefined` when we try to access an specific query param during start up.

Reviewer 2: **thought:** this paragraph mostly rehashes the previous two, maybe you can merge the last three paragraphs and make them slightly simpler? This Angular quirk is not necessarily the core interesting part of this article and I don't think we need to go quite that deep into it.

This whole situation affects us in the initial value we provide to our `queryParamValue` signal. If we use the `get` method from the `QueryParamService` in components that are loaded during the initial navigation, then our `queryParamValue` signal will be initialized to `undefined`. This is technically not a big deal since it is connected to the `activatedRoute.queryParams` observable. So as soon as Angular emits the actual value of the query param, then the signal will be updated accordingly.

The issue with the empty UI is related to another not so intuitive behavior of Angular's router: Cancelling concurrent navigations. Maybe this surprises you but when we do `router.navigate`, this is actually an asynchronous operation, it returns a `Promise`, which is resolved whenever the navigation is completed.

> The problem is when we start or trigger a `router.navigate` when there's a navigation that is still on progress. When Angular detects this, it will **cancel the current navigation** and execute the new one instead, completely ignoring the previous one. How is this related to the empty UI problem? Let's take a look:

Reviewer 2: ```suggestion
The problem occurs when we trigger `router.navigate` while a navigation is still in progress. When Angular detects this, it will **cancel the current navigation** and execute the new one instead, completely ignoring the previous one. How is this related to the empty UI problem? Let's take a look:
```

As soon as this navigation finishes Angular could emit the actual query param values through the `activatedRoute.queryParams` observable, but there are no query params in `http://localhost:4200/`.

The reason why there are no query params in this new navigation is in the `router.navigate` we are triggering in the effect:

```ts
  queryParams: { [queryParamName]: queryParamValue() },
```

In this example `queryParamName` is `"firstName"` and `queryParamValue()` is initially `undefined` for what we mentioned a bit above (you could think that we are accessing `activatedRoute.snapshot.queryParams["firstName"]` during the period of time in which the query params object is empty `{}`). 

So the effect is triggering a navigation that has this information: "navigate to `http://localhost:4200?firstName=undefined`", since the value of `firstName` is `undefined` it's just removed from the query params. And notice that there's no information about the other query params at all.

> We could argue that we are indicating Angular to `merge` the query params in that `router.navigate` call. So no query params should be removed from the browser url.

Reviewer 2: **thought:** I think the whole bit starting with "Let's try to understand …" on line 471 is a bit overkill, it's overly detailed and repetitive. Readers that came for Signals will jump off here. Try to reduce it to the absolute minimum explaining the problem but give readers some credit :D

  private allQueryParams = toSignal(this.activatedRoute.queryParams);

  constructor() {
    // Notice that we log navigation events here
    this.router.events.subscribe((e) => console.log(e));
  }

  // Same as before
  get(queryParamName: string) {
    const activatedRouteValue = computed(() => this.allQueryParams()?.[queryParamName]);
    const queryParamValue = signal(activatedRouteValue());

    effect(() => queryParamValue.set(activatedRouteValue()), { allowSignalWrites: true });

    effect(() =>
      this.router.navigate([], {
        queryParams: { [queryParamName]: queryParamValue() },
        queryParamsHandling: "merge",
      })
    );

    return queryParamValue;
  }
}
```

```ts
@Component({
  selector: "app-root",
  standalone: true,
  template: `
    <h1>Signals Demo</h1>
    <p>Your first name is: {{ firstName() }}</p>
  `,
})
export class AppComponent {
  private queryParamService = inject(QueryParamService);

  // Notice we only use the .get() method once
  firstName = this.queryParamService.get("firstName");
}
```

![Angular cancelling navigations](https://cdn.hashnode.com/res/hashnode/image/upload/v1710594428122/YXorJBf0j.gif?auto=format)

> Notice that when we visit the url in the browser, Angular triggers an initial navigation on start up, this initial navigation contains the query params actual values, those that are initially provided in the browser url. But, as soon as we execute the second effect inside the `queryParamService.get` method, a new empty navigation is triggered, which cancels the initial one, losing with it all query params data.

Reviewer 2: **thought:** again, this is just digging into the same problem for the 3rd(?) time, let's go to the solution quicker

I could see this be a separate article to be honest, something like "Why initial Angular Query Params are missing" or "How Angular Active Route works" or something like that. It's an interesting topic, but makes this article lose focus.

For this reason, we somehow need to **wait** for the current navigation to complete in order to execute the one that we have in the second effect. Fortunately, we can take advantage of `async/await` when dealing with signals, we just have to be careful so we don't get unexpected situations, pretty similar to what we discussed with `allowSignalWrites: true`.

This is the final version of the service, including logic to wait for the current navigation to complete:

```ts
@Injectable({ providedIn: "root" })
export class QueryParamService {
  private activatedRoute = inject(ActivatedRoute);
  private router = inject(Router);

  private allQueryParams = toSignal(this.activatedRoute.queryParams);

  get(queryParamName: string) {
    const activatedRouteValue = computed(() => this.allQueryParams()?.[queryParamName]);
    const queryParamValue = signal(activatedRouteValue());

    effect(() => queryParamValue.set(activatedRouteValue()), { allowSignalWrites: true });

    effect(async () => {
      const newValueToPushInRoute = queryParamValue();

      if (!this.router.navigated) return;

>       if (this.isRouterNavigating()) await this.waitUntilRouterIsIdle();

Reviewer 2: **question:** what happens if `queryParamValue` has multiple new value while this Promise hasn't resolve yet. Will that start the async effect multiple times and consequently execute `this.router.navigate` multiple times? Could that lead to race conditions?

    <hr />

    <h2>Using Angular Router</h2>
    <button (click)="becomeHarryPotter()">Become Harry Potter 🪄</button>
    <button (click)="getYounger()">Get Younger 👶</button>
  `,
})
export class AppComponent {
  private queryParamService = inject(QueryParamService);
  private router = inject(Router);
  private activatedRoute = inject(ActivatedRoute);

  firstName = this.queryParamService.get('firstName');
  lastName = this.queryParamService.get('lastName');
  age = this.queryParamService.get('age');

  fullName = computed(
    () => `${this.firstName() ?? ''} ${this.lastName() ?? ''}`
  );

  // ---- using signals
  anonymise() {
    this.firstName.set(undefined);
    this.lastName.set(undefined);
  }

  becomeJackSparrow() {
    this.firstName.set('Jack');
    this.lastName.set('Sparrow');
  }

  getOlder() {
    this.age.set(+this.age() + 5);
  }

  // ---- using Angular router
  becomeHarryPotter() {
    const queryParams = { firstName: 'Harry', lastName: 'Potter' };
    this.router.navigate([], { queryParams, queryParamsHandling: 'merge' });
  }

  getYounger() {
    const queryParams = {
      age: +this.activatedRoute.snapshot.queryParams['age'] - 5,
    };

>     this.router.navigate([], { queryParams, queryParamsHandling: 'merge' });

Reviewer 2: **question:** why aren't these methods using the Signal setters but still go directly through the router?
